File: generators/en16_generator.py
from datetime import datetime
from io import BytesIO
from PyPDF2 import PdfReader, PdfWriter
from reportlab.pdfgen import canvas
from reportlab.lib.pagesizes import letter


class EN16Generator:
    def __init__(self, template_path="templates/en-16.pdf"):
        self.template_path = template_path

    def generate(self, client_record: dict) -> tuple[str, BytesIO]:
        raw_name = client_record["fields"].get("Name", "")
        try:
            case_id = client_record["fields"].get("Case ID", "")
            last_first = raw_name.split(" - ")[0]
            last, first = [s.strip() for s in last_first.split(",")]
            client_name = f"{first} {last}"
        except ValueError:
            raise ValueError(
                "Client name format invalid. Expected 'Last, First - ####'"
            )

        current_date = datetime.now().strftime("%m.%d.%y")
        filename = f"EN16_{first[0]}.{last}_{current_date}.pdf"

        pdf_bytes = self.draw_pdf(self.template_path, client_name, case_id)
        return filename, pdf_bytes

    def draw_pdf(self, template_path: str, client_name: str, case_id: str) -> BytesIO:
        base_pdf = PdfReader(template_path)
        writer = PdfWriter()

        # First page: name + case ID
        overlay_page_1 = self._make_overlay_page(
            client_name=client_name,
            case_id=case_id,
            draw_date=False,
        )
        first_page = base_pdf.pages[0]
        first_page.merge_page(overlay_page_1)
        writer.add_page(first_page)

        # Middle pages (no overlay)
        for i in range(1, len(base_pdf.pages) - 1):
            writer.add_page(base_pdf.pages[i])

        # Last page: current date only
        overlay_last = self._make_overlay_page(
            client_name=None,
            case_id=None,
            draw_date=True,
        )
        last_page = base_pdf.pages[-1]
        last_page.merge_page(overlay_last)
        writer.add_page(last_page)

        result = BytesIO()
        writer.write(result)
        result.seek(0)
        return result

    def _make_overlay_page(self, client_name=None, case_id=None, draw_date=False):
        overlay_buffer = BytesIO()
        overlay = canvas.Canvas(overlay_buffer, pagesize=letter)
        overlay.setFont("Helvetica", 12)
        if client_name and case_id:
            overlay.drawString(355, 695, client_name)
            overlay.drawString(355, 710, case_id)
        if draw_date:
            overlay.drawString(250, 227, datetime.now().strftime("%m/%d/%Y"))
        overlay.save()
        overlay_buffer.seek(0)
        return PdfReader(overlay_buffer).pages[0]

File: web/src/lib/generators/en16-generator.ts
/**
 * EN-16 Form Generator
 * Converts Python en16_generator.py to TypeScript
 */

import { BaseGenerator } from './base-generator';
import { GeneratorResult } from './types';
import { formatDateMMDDYY, formatDateMMDDYYYY } from './utils/formatters';
import { StandardFonts } from 'pdf-lib';

export interface EN16FormData {
  claimant: string;
  case_id: string;
}

export class EN16Generator extends BaseGenerator {
  constructor() {
    super('en-16.pdf');
  }

  async generate(
    claimant: string,
    caseId: string
  ): Promise<GeneratorResult> {
    const nameParts = claimant.trim().split(' ');
    if (nameParts.length < 2) {
      throw new Error(
        'Claimant name format invalid. Must include at least first and last name.'
      );
    }

    const first = nameParts[0];
    const last = nameParts[nameParts.length - 1];

    // Generate filename
    const currentDate = formatDateMMDDYY();
    const filename = `EN16_${first[0]}.${last}_${currentDate}.pdf`;

    // Load template
    const pdfDoc = await this.loadTemplate();
    const font = await pdfDoc.embedFont(StandardFonts.Helvetica);
    const pages = pdfDoc.getPages();

    // First page: name + case ID
    if (pages.length > 0) {
      const firstPage = pages[0];
      firstPage.setFont(font);
      firstPage.setFontSize(12);

      this.drawText(firstPage, claimant, { x: 355, y: 695, size: 12 });
      this.drawText(firstPage, caseId, { x: 355, y: 710, size: 12 });
    }

    // Middle pages: no overlay (skip)

    // Last page: current date only
    if (pages.length > 0) {
      const lastPage = pages[pages.length - 1];
      lastPage.setFont(font);
      lastPage.setFontSize(12);

      const currentDateFormatted = formatDateMMDDYYYY();
      this.drawText(lastPage, currentDateFormatted, { x: 250, y: 227, size: 12 });
    }

    // Save PDF
    const pdfBytes = await pdfDoc.save();

    return {
      filename,
      pdfBytes: Buffer.from(pdfBytes),
    };
  }
}
